### ISSUE TITLE
[no-magic-numbers] false positives everywhere, totally broken

### ISSUE THREAD

**Issue by David-Else** [typescript-eslint/typescript-eslint#766]
**Repro**

```JSON
"@typescript-eslint/no-magic-numbers": "warn",
```

```TS
function isHiscores(obj: unknown): obj is Hiscores {
  return (
    obj instanceof Array &&
    typeof obj[0] === 'object' &&  // <<<<<<<
    typeof obj[0].name === 'string' && // <<<<<<<
    typeof obj[0].score === 'number' // <<<<<<<
  );
}

//

export abstract class GameObject {
  public rotation = 0;
  //   protected scale: Vector2 = [0, 0];
  protected abstract position: Vector2 = [0, 0];

  public get x(): number {
    return this.position[0]; // <<<<<<<
  }
  public get y(): number {
    return this.position[1]; // <<<<<<<
  }
}

//

  if (!context.entities.hero.lives) {
    context.entities.hero = instantiate(Hero, 1, {  // <<<<<<<
      position: [context.ctx.canvas.width / 2, context.ctx.canvas.height / 2],     });
  }  // <<<<<<<

```

**Expected Result**
nothing
**Actual Result**
```
/home/david/sites/neweslintsetup/src/http-service.ts:40:16: No magic number: 0. [Warning/@typescript-eslint/no-magic-numbers]
/home/david/sites/neweslintsetup/src/http-service.ts:41:16: No magic number: 0. [Warning/@typescript-eslint/no-magic-numbers]
/home/david/sites/neweslintsetup/src/http-service.ts:42:16: No magic number: 0. [Warning/@typescript-eslint/no-magic-numbers]

/home/david/sites/neweslintsetup/src/entities/base-classes.ts:20:26: No magic number: 0. [Warning/@typescript-eslint/no-magic-numbers]
/home/david/sites/neweslintsetup/src/entities/base-classes.ts:23:26: No magic number: 1. [Warning/@typescript-eslint/no-magic-numbers]

/home/david/sites/neweslintsetup/src/states/level-one.ts:23:49: No magic number: 1. [Warning/@typescript-eslint/no-magic-numbers]
/home/david/sites/neweslintsetup/src/states/level-one.ts:24:47: No magic number: 2. [Warning/@typescript-eslint/no-magic-numbers]
/home/david/sites/neweslintsetup/src/states/level-one.ts:24:78: No magic number: 2. [Warning/@typescript-eslint/no-magic-numbers]
```
**Additional Info**

**Versions**
```
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^1.13.0",
    "@typescript-eslint/parser": "^1.13.0",
    "eslint": "^6.1.0",
    "eslint-config-prettier": "^6.0.0",
    "typescript": "^3.5.3"
  }
```
Here is my full config:
```
{
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "prettier",
    "prettier/@typescript-eslint"
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "plugins": ["@typescript-eslint"],
  "rules": {
    /**
     * ========================================================================
     * Airbnb JS Style Guide rules extending eslint:recommended
     * ========================================================================
     */
    // require the use of === and !==
    "eqeqeq": ["error", "always", { "null": "ignore" }],
    // disallow else after a return in an if
    "no-else-return": ["error", { "allowElseIf": false }],
    // enforce that class methods use "this"
    "class-methods-use-this": [
      "error",
      {
        "exceptMethods": []
      }
    ],
    // disallow redundant return; keywords
    "no-useless-return": "error",
    // require let or const instead of var
    "no-var": "error",
    // require method and property shorthand syntax for object literals
    "object-shorthand": [
      "error",
      "always",
      {
        "ignoreConstructors": false,
        "avoidQuotes": true
      }
    ],
    // suggest using arrow functions as callbacks
    "prefer-arrow-callback": [
      "error",
      {
        "allowNamedFunctions": false,
        "allowUnboundThis": true
      }
    ],
    // suggest using const declaration for vars never modified after declared
    "prefer-const": [
      "error",
      {
        "destructuring": "any",
        "ignoreReadBeforeAssign": true
      }
    ],
    // Prefer destructuring from arrays and objects
    "prefer-destructuring": [
      "error",
      {
        "VariableDeclarator": {
          "array": false,
          "object": true
        },
        "AssignmentExpression": {
          "array": true,
          "object": true
        }
      },
      {
        "enforceForRenamedProperties": false
      }
    ],
    // suggest using template literals instead of string concatenation
    "prefer-template": "error",
    /**
     * ========================================================================
     * ESLint Plugin TypeScript rules extending @typescript-eslint/recommended
     * ========================================================================
     */
    // disallow unused variables
    "@typescript-eslint/no-unused-vars": "off", // "warn" for production
    // disallow the use of parameter properties in class constructors
    "@typescript-eslint/no-parameter-properties": "off",
    // warns if a type assertion does not change the type of an expression
    "@typescript-eslint/no-unnecessary-type-assertion": "warn",
    // disallows magic numbers
    "@typescript-eslint/no-magic-numbers": "warn",
    // require a consistent member declaration order
    "@typescript-eslint/member-ordering": "warn",
    // Warns for any two overloads that could be unified into one by using a union or an optional/rest parameter
    "@typescript-eslint/unified-signatures": "warn",
    // requires Promise-like values to be handled appropriately
    "@typescript-eslint/no-floating-promises": "warn",
    // requires any function or method that returns a Promise to be marked async
    "@typescript-eslint/promise-function-async": "warn",
    // disallow async functions which have no await expression
    "@typescript-eslint/require-await": "warn",
    // avoid using promises in places not designed to handle them
    "@typescript-eslint/no-misused-promises": "warn",
    // require explicit return types on functions and class methods
    "@typescript-eslint/explicit-function-return-type": [
      "warn",
      {
        "allowExpressions": true
      }
    ]
  },
  "env": {
    "browser": true,
    "node": true,
    "mocha": true
  },
  "globals": {
    "assert": true
  }
}
```
** edit same results with `"no-magic-numbers": "off",` added, but it was never on.

**Comment by bradzacher** [comment]
This is just how the base rule works with no config: [eslint repl](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tbWFnaWMtbnVtYmVyczogW1wiZXJyb3JcIl0qL1xuZnVuY3Rpb24gaXNIaXNjb3JlcyhvYmopIHtcbiAgcmV0dXJuIChcbiAgICBvYmogaW5zdGFuY2VvZiBBcnJheSAmJlxuICAgIHR5cGVvZiBvYmpbMF0gPT09ICdvYmplY3QnICYmICAvLyA8PDw8PDw8XG4gICAgdHlwZW9mIG9ialswXS5uYW1lID09PSAnc3RyaW5nJyAmJiAvLyA8PDw8PDw8XG4gICAgdHlwZW9mIG9ialswXS5zY29yZSA9PT0gJ251bWJlcicgLy8gPDw8PDw8PFxuICApO1xufVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo5LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6e30sImVudiI6eyJjb21tb25qcyI6ZmFsc2UsIm5vZGUiOmZhbHNlfX19)

If you want to not catch this case, please see the options in the rule docs:
https://eslint.org/docs/rules/no-magic-numbers

The `ignoreArrayIndexes` option will handle the use case you want.

**Comment by David-Else** [comment]
Thanks! I did not know that certain `@typescript-eslint/xxx` rules that were extensions of eslint rules could use their options! I changed the rule to this and now things are a lot more sane!
```
    "@typescript-eslint/no-magic-numbers": [
      "warn",
      {
        "ignoreArrayIndexes": true,
        "ignore": [0, 1]
      }
    ],
```
Perhaps consider pointing out these options in the typescript-eslint docs as they are vital and easy to miss.

**Comment by bradzacher** [comment]
I believe it should be documented within the rule's documentation:

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-magic-numbers.md#rule-details

> The `@typescript-eslint/no-magic-numbers` rule extends the `no-magic-numbers` rule from ESLint core, and adds support for handling Typescript specific code that would otherwise trigger the rule.
>
> See the [ESLint documentation](https://eslint.org/docs/rules/no-magic-numbers) for more details on the `no-magic-numbers` rule.

**Comment by David-Else** [comment]
I saw that, but my problem was I did not understand that 

> extends the xxx rule 

meant that you could use all the options of the original rule. Not knowing the inner working of the extension I just assumed that the rules were separate things. Reading it now I can see that 'extends' probably would meant this, but it was not obvious.

Maybe it is just me being dumb, but if you could be more explicit about this somewhere then it might help someone avoid what happened to me. Maybe a little icon next to the rules that extend eslint rules, and a note saying you can use all the original options. Just an idea. Cheers!

**Comment by bradzacher** [comment]
There is also this line in the next section, titled "Rule Changes":

> In addition to the options supported by the no-magic-numbers rule in ESLint core, the rule adds the following options

We could potentially mark this down more clearly.

**Comment by David-Else** [comment]
While I have your attention, how about a rule that says ignore any numbers in arrays? My game uses a Vector2 class. All over the code are coordinates like:
```
public position: Vector2 = [200, 200];
```
and they get identified as magic numbers. Am I missing something obvious? This kind of thing must be very common in many types of program.

**Comment by bradzacher** [comment]
unfortunately those are by definition magic numbers.
the point of the no-magic-numbers rule is so that you don't use number values in your code, you should refactor them to constant variables which are intended to make your code easier to understand:

```ts
const INITIAL_COORDINATE = 200;

public position: Vector2 = [INITIAL_COORDINATE, INITIAL_COORDINATE];
```

I would probably just suggest not using the rule, considering your intention is to use so-called magic numbers?

**Comment by David-Else** [comment]
I will see what the code looks like with your INITIAL_COORDINATE idea, and think hard about if I will use the magic number rule after all, thanks for your help!
